src/HexagonOptimize.cpp

```diff
@@ -2046,8 +2049,19 @@ class VtmpyGenerator : public IRGraphMutator {
                         if (vtmpy_indices[i]) {
                             continue;
                         }
-                        Expr mpy_a = lossless_cast(op->type, mpys[i].first);
-                        Expr mpy_b = lossless_cast(op->type, mpys[i].second);
+                        // We put expressions in mpys after un-broadcasting them. So, first broadcast
+                        // then call lossless_cast.
+                        Expr a = mpys[i].first;
+                        Expr b = mpys[i].second;
+                        int lanes = op->type.lanes();
+
+                        if (a.type().lanes() < lanes)
```

Comment:
I think this should be a.type().lanes() == 1. Broadcast only makes sense for lanes == 1, and maybe there are some ops (shuffles) that would pass this check but break when broadcasted.

Same for b below.

Reply:
Nit: a.type().is_scalar()

Reply:
Done.